platform/default/layer_manager.cpp

```diff
@@ -0,0 +1,61 @@
+#include <mbgl/style/layer.hpp>
+#include <mbgl/style/layers/symbol_layer.hpp>
+#include <mbgl/style/layers/background_layer.hpp>
+#include <mbgl/style/layers/circle_layer.hpp>
+#include <mbgl/style/layers/fill_extrusion_layer.hpp>
+#include <mbgl/style/layers/fill_layer.hpp>
+#include <mbgl/style/layers/heatmap_layer.hpp>
+#include <mbgl/style/layers/hillshade_layer.hpp>
+#include <mbgl/style/layers/line_layer.hpp>
+#include <mbgl/style/layers/raster_layer.hpp>
+#include <mbgl/style/layers/symbol_layer.hpp>
+
+#include <memory>
+#include <vector>
+
+namespace mbgl {
+namespace style {
+ 
+class LayerManagerBase : public LayerManager {
+public:
+    LayerManagerBase();
+private:
+    // LayerManager overrides.
+    std::unique_ptr<Layer> createLayer(const std::string& type, const std::string& id, const conversion::Convertible& value, conversion::Error& error) final;
+    std::vector<std::unique_ptr<LayerFactory>> factories;
+};
+
+LayerManagerBase::LayerManagerBase() {
+    factories.emplace_back(std::unique_ptr<LayerFactory>(new FillLayerFactory));
+    factories.emplace_back(std::unique_ptr<LayerFactory>(new LineLayerFactory));
+    factories.emplace_back(std::unique_ptr<LayerFactory>(new CircleLayerFactory));
+    factories.emplace_back(std::unique_ptr<LayerFactory>(new SymbolLayerFactory));
+    factories.emplace_back(std::unique_ptr<LayerFactory>(new RasterLayerFactory));
+    factories.emplace_back(std::unique_ptr<LayerFactory>(new BackgroundLayerFactory));
+    factories.emplace_back(std::unique_ptr<LayerFactory>(new HillshadeLayerFactory));
+    factories.emplace_back(std::unique_ptr<LayerFactory>(new FillExtrusionLayerFactory));
+    factories.emplace_back(std::unique_ptr<LayerFactory>(new HeatmapLayerFactory));
+}
+
+std::unique_ptr<Layer> LayerManagerBase::createLayer(const std::string& type, const std::string& id, const conversion::Convertible& value, conversion::Error& error) {
+    for (const auto& factory: factories) {
```

Comment:
It would enforce a type name for the custom layer, which I wanted to obviate.
Otherwise, it would be either part of the style or some special logic for custom layers inside the layer manager.

Reply:
custom layers do not support any type (by "type" I mean type in a JSON style document) unless the opposite will be intended. So, for now we can just define it as:
```
bool CustomLayerFactory::supportsType(const std::string&) const {
    return false;
}
```